x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequestSender.java

```diff
@@ -97,12 +107,35 @@ final void startComputeOnDataNodes(
                         nodePermits.putIfAbsent(node, new Semaphore(1));
                     }
                 }
-                pendingShardIds.addAll(targetShards.shards.keySet());
+                pendingShardIds.addAll(order(targetShards));
                 trySendingRequestsForPendingShards(targetShards, computeListener);
             }
         }, listener::onFailure));
     }
 
+    private static List<ShardId> order(TargetShards targetShards) {
+        var ordered = new ArrayList<>(targetShards.shards.keySet());
+        ordered.sort(Comparator.comparingInt(shardId -> nodeOrder(targetShards.getShard(shardId).remainingNodes)));
+        return ordered;
+    }
+
+    private static int nodeOrder(List<DiscoveryNode> nodes) {
+        if (nodes.isEmpty()) {
+            return Integer.MAX_VALUE;
+        }
+        // assumes all shard nodes have same roles
+        var node = nodes.getFirst();
+        if (node.hasRole(DiscoveryNodeRole.DATA_CONTENT_NODE_ROLE.roleName()) || node.hasRole(DiscoveryNodeRole.DATA_ROLE.roleName())) {
+            return 0;
+        }
+        for (int i = 0; i < NODE_QUERY_ORDER.length; i++) {
+            if (node.hasRole(NODE_QUERY_ORDER[i])) {
+                return i;
+            }
+        }
+        return NODE_QUERY_ORDER.length;
+    }
```

Comment:
I have a sneaking suspicion this is going to be expensive enough to calculate that we should do it once for each one and stuff it in record or something.

Reply:
Just in case, this is executed only once when adding requests to the queue first time

Reply:
Added a map to reuse computation result https://github.com/elastic/elasticsearch/pull/122928/commits/b6435af32a6a02d5837a39fc5ed6feecfdc7fe0e